bundles/org.openhab.binding.tsmart/src/main/java/org/openhab/binding/tsmart/internal/TSmartDiscoveryService.java

```diff
@@ -0,0 +1,93 @@
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.tsmart.internal;
+
+import static org.openhab.binding.tsmart.internal.TSmartBindingConstants.THING_TYPE_T_SMART;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.core.config.discovery.AbstractDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResult;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.config.discovery.DiscoveryService;
+import org.openhab.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link TSmartDiscoveryService} is used to discover T-Smart devices that are connected to the local network.
+ *
+ * @author James Melville - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = DiscoveryService.class, configurationPid = "discovery.tsmart")
+public class TSmartDiscoveryService extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(TSmartDiscoveryService.class);
+
+    public TSmartDiscoveryService() {
+        super(Set.of(THING_TYPE_T_SMART), 60, false);
+    }
+
+    /**
+     * Start Discovery of T-Smart devices
+     */
+    @Override
+    protected void startScan() {
+        TSmartUDPListener.startDiscovery(this);
+
+        try {
+            logger.debug("Started discovery by sending UDP packet to broadcast address");
+            new TSmartUDPUtils().sendUDPPacket(InetAddress.getByName("255.255.255.255"),
+                    new byte[] { (byte) 0x01, (byte) 0x00, (byte) 0x00 });
+        } catch (UnknownHostException e) {
+            logger.debug("Unknown host: {}", e.getMessage());
+        }
+    }
+
+    /**
+     * End Discovery of T-Smart devices
+     */
+    @Override
+    protected void stopScan() {
+        super.stopScan();
+        TSmartUDPListener.stopDiscovery();
+    }
+
+    /**
+     * Receive the UDP packet responses for the discovery broadcast packet.
+     * Parse these and route to the Inbox
+     *
+     * @param addr Network address of device
+     * @param buffer Packet bytes
+     */
+    public void handleDiscoveryResponse(InetAddress addr, byte[] buffer) {
+        String name = new String(Arrays.copyOfRange(buffer, 9, 20));
+        String id = String.format("%02X%02X%02X%02X", buffer[8], buffer[7], buffer[6], buffer[5]);
+
+        Map<String, Object> properties = new HashMap<>(3);
+        properties.put("hostname", addr.getHostAddress());
+        properties.put("id", id);
+        properties.put("name", name);
```

Comment:
Consider adding constants for these keys.

I recognize `hostname` as a configuration parameter. Are `id` and `name` properties? Will they also be populated when creating the thing manually, or only through discovery?

Reply:
Sorry, I wrote this code a long time ago and my mind is not fresh! Now I look into it discovery isn't working for me, so will debug my environment and get it working before any resubmission.

- Yes, hostname is the configuration parameter.
- Name is user entered in the device's app, so I figured it should be the default label for a new device.
- Id is fixed and derived from the MAC address, hence why I used it as the representation property. But I assume that wouldn't work properly because I'm not setting the property in normal usage.

I think the best fix for this is setting the name and id properties in runtime, outside of the discovery process.

Reply:
Wrt label, it's already provided below with `.withLabel(name)`, so as a property I'm not sure if relevant.

Mostly it's relevant to provide properties to feel the configuration parameters where this is possible. For the rest, you are right that it should be populated by the handler itself so that it will also work for manual configuration and even file-based configuration.